[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96718

**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -901,6 +931,57 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+      "qlinear (ONEDNN): data type of input should be int8 (char).");
+  TORCH_CHECK(
+      inv_weight_scales.scalar_type() == c10::ScalarType::Float, "weight scales should be dtype c10::ScalarType::Float.");
+  ideep::scale_t weights_scales(
+      (float*)inv_weight_scales.data_ptr(), (float*)inv_weight_scales.data_ptr() + inv_weight_scales.numel());
```
I think we need a zero-copy solution to avoid overhead here.

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -207,6 +208,50 @@ c10::intrusive_ptr<LinearPackedParamsBase> PackedLinearWeightFp16::prepack(
+  return wei;
+}
+
+inline ideep::tensor pack_bias(
+    const at::Tensor& b,
+    int64_t oc,
+    ideep::scale_t weight_scales = ideep::scale_t(),
+    double input_scale = 1,
+    int64_t input_zero_point = 0) {
```
Let's design for onednn 3.x in which bias packing is not needed. Then, the code here can be simplified further.

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][ONEDNN] add a qlinear path that accepts int8 tensor from CPU backend for future use in Inductor"


**Summary**
As the title

**Test plan**
python test/test_quantization.py TestQuantizedLinear.test_qlinear_with_cpu_tensor_onednn

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -4035,6 +4035,74 @@ def test_qlinear_tanh(self):
+        - output: linear result as an int8 tensor from CPU backend
+        These ops are used for Inductor quantization
+        '''
+        qlinear_prepack = torch.ops.quantized.linear_prepack_mkldnn
+        qlinear = torch.ops.quantized.linear_with_mkldnn_weight_bias
```
can we align the names, e.g. rename mkldnn to onednn?

Xia-Weiwen replied:
Thanks for the suggestion. I will have it fixed after the onednn update and rebase.

kimishpatel commented:
Can you more on the relation of these changes to inductor?

Xia-Weiwen commented:
> Can you more on the relation of these changes to inductor?

Sure. We are working on quantization PT2E and inductor will be the backend for x86 CPU for the new quantization path. These new ops will be called by Inductor for quantization.

Xia-Weiwen commented:
Close this PR as we have a new one here: https://github.com/pytorch/pytorch/pull/105818, which does exactly the same thing.
